python/ray/dag/compiled_dag_node.py

```diff
@@ -150,54 +138,58 @@ def _wrap_exception(exc):
     return wrapped
 
 
-def _exec_task(self, task: "ExecutableTask", idx: int) -> bool:
+def _select_next_nodes(
+    actor_to_candidates: Dict["ray.actor.ActorHandle", List[DAGOperationGraphNode]],
+    graph: Dict[int, Dict[DAGNodeOperationType, DAGOperationGraphNode]],
+):
     """
-    Execute the task.
+    This function selects the next nodes for topological sort to generate execution
+    schedule. If there are multiple DAGOperationGraphNodes with zero in-degree,
+    select nodes based on the following rules:
+
+    #1  If the nodes are not NCCL write nodes, select the one with the smallest
+        `bind_index`. If there are multiple candidate nodes with the smallest
+        `bind_index` of the actors that they belong to, any one of them is
+        acceptable. For the implementation details, we maintain a priority queue
+        for each actor, where the head of the priority queue is the node with the
+        smallest `bind_index`.
+    #2  If #1 cannot be satisfied, it means that all candidate nodes are
+        NCCL write nodes. In this case, select the one that is the head of the
+        priority queue and its downstream nodes, regardless of whether the
+        downstream nodes are heads of their priority queues or not.
+
+    This function may return multiple nodes if they are NCCL nodes. In that case,
+    this function only removes the NCCL write node, which is also the head of a
+    priority queue. Other nodes will be removed in the following iterations.
+    Additionally, visited_nodes ensures that the same node will not be scheduled
+    more than once.
+
     Args:
-        task: The task to execute.
-        idx: The index of the task in the list of tasks of the actor.
-    Returns:
-        True if we are done executing all tasks of this actor, False otherwise.
+        actor_to_candidates: A dictionary mapping an actor handle to a list of
+            candidate nodes with zero in-degree.
```

Comment:
is there any assumption in the list of candidate nodes? Like the smaller index is higher priority?

Reply:
`candidates` is a list which is maintained as a priority queue, so the head of the queue, i.e., `candidates[0]`, is
the node with the smallest `bind_index`.

Reply:
Updated https://github.com/ray-project/ray/pull/46911/commits/ed5836d3684a81ad7375b5177e36a252be7e32e0